File: presupuestos-backend/src/main/java/com/neria/presupuestos/model/entity/Formula.java
package com.neria.presupuestos.model.entity;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import jakarta.persistence.PrePersist;
import jakarta.persistence.Table;
import lombok.Getter;
import lombok.Setter;
import org.hibernate.annotations.Filter;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.UUID;

@Getter
@Setter
@Entity
@Table(name = "presupuestos_formulas")
@Filter(name = "tenantFilter", condition = "tenant_id = :tenantId")
public class Formula {
    @Id
    @Column(length = 36, columnDefinition = "CHAR(36)")
    private String id;

    @Column(name = "tenant_id", length = 36, columnDefinition = "CHAR(36)")
    private String tenantId;

    @Column(name = "sector_id", length = 36, columnDefinition = "CHAR(36)")
    private String sectorId;

    @Column(name = "product_id", length = 36, columnDefinition = "CHAR(36)")
    private String productId;

    @Column(nullable = false)
    private String name;

    @Column(columnDefinition = "TEXT")
    private String description;

    @Column(name = "base_price")
    private BigDecimal basePrice;

    @Column(name = "unit_price")
    private BigDecimal unitPrice;

    @Column(nullable = false)
    private boolean active = true;

    @Column(name = "created_at")
    private LocalDateTime createdAt;

    @PrePersist
    void onCreate() {
        if (id == null || id.isBlank()) {
            id = UUID.randomUUID().toString();
        }
        if (createdAt == null) {
            createdAt = LocalDateTime.now();
        }
    }
}

File: presupuestos-frontend/src/types/formula.ts
export interface Formula {
  id: string;
  tenantId?: string;
  sectorId?: string;
  sectorName?: string;
  productId?: string;
  productName?: string;
  name: string;
  description?: string;
  basePrice?: number;
  unitPrice?: number;
  active: boolean;
  createdAt?: string;
}

export interface FormulaCreateRequest {
  sectorId: string;
  productId?: string;
  name: string;
  description?: string;
  basePrice?: number;
  unitPrice?: number;
  active?: boolean;
}

export interface FormulaUpdateRequest {
  sectorId?: string;
  productId?: string;
  name?: string;
  description?: string;
  basePrice?: number;
  unitPrice?: number;
  active?: boolean;
}
